File: 96.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.count = 0       # number of words that pass through this node
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
            node.count += 1  # increment count for prefix
        node.isEnd = True

    def countPrefix(self, prefix):
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return 0
            node = node.children[ch]
        return node.count


# Example
trie = Trie()
words = ["apple", "app", "apt", "bat", "ball"]
for w in words:
    trie.insert(w)

print(trie.countPrefix("ap"))   # 3
print(trie.countPrefix("ba"))   # 2
print(trie.countPrefix("app"))  # 2

File: 96.cs
using System.Collections.Generic;

public class TrieNode
{
    public Dictionary<char, TrieNode> Children = new Dictionary<char, TrieNode>();
    public int Count = 0;     // number of words that pass through
    public bool IsEnd = false;
}

public class Trie
{
    private TrieNode root = new TrieNode();

    public void Insert(string word)
    {
        TrieNode node = root;
        foreach (char ch in word)
        {
            if (!node.Children.ContainsKey(ch))
                node.Children[ch] = new TrieNode();

            node = node.Children[ch];
            node.Count++;         // increment prefix count
        }
        node.IsEnd = true;
    }

    public int CountPrefix(string prefix)
    {
        TrieNode node = root;
        foreach (char ch in prefix)
        {
            if (!node.Children.ContainsKey(ch))
                return 0;
            node = node.Children[ch];
        }
        return node.Count;
    }
}

// Example:
/*
Trie trie = new Trie();
string[] words = { "apple", "app", "apt", "bat", "ball" };
foreach (var w in words) trie.Insert(w);

Console.WriteLine(trie.CountPrefix("ap"));   // 3
Console.WriteLine(trie.CountPrefix("ba"));   // 2
Console.WriteLine(trie.CountPrefix("app"));  // 2
*/
